app/server.py

```diff
@@ -22,6 +22,504 @@ def allRecords():
+        # Add to userdb
+        db_users.users.insert(new_user)
+        return "User: " + username + " has been created with Pass: " + hash_password
+
+
+# Login user, authorization
+@app.route('/login/<username>/<password>')
+def login(username, password):
+    user = db_users.users.find_one({'username': username})
+    if user:
+        if user['token'] == '':
+            # Determine if password is correct
+            if hash_pass(password) == user['password']:
+                token = generate_token(username)
+                print token
+                # TODO: Send Token Here
+                # Store the token in userdb
+                db_users.users.update_one({'username': username}, {'$set': {'token': token}})
+                return "User: " + username + " has logged in"
+            else:
+                return "Error: Given password does not match with user: " + username
+        else:
+            if is_valid_token(user['token'], user['username']):
+                return "Error: User: " + username + " is already logged in"
+            else:
+                # Reset the token in userdb
+                db_users.users.update_one({'username' : username}, { '$set': {'token': ""} })
+                return "Error: Invalid token received, reset token"
+    else:  # User not found
+        return "Error: User: " + username + " not found"
+
+
+# Logout user, authorization
+# TODO: Instead of Username, receive Token
+@app.route('/logout/<username>')
+def logout(username):
+    user = db_users.users.find_one({'username': username})
+    if user:
+        if user['token'] == '':
+            return "Error: User: " + username + " is not logged in"
+        else:
+            if is_valid_token(user['token'], user['username']):
+                # Reset the token in userdb
+                db_users.users.update_one({'username': username}, {'$set': {'token': ""}})
+                return "User: " + username + " has been logged out"
+            else:
+                # Reset the token in userdb
+                db_users.users.update_one({'username': username}, {'$set': {'token': ""}})
+                return "Error: Invalid token received, reset token"
+    else:  # User not found
+        return "Error: User: " + username + " not found"
+
+
+# Hash a given password using the sha1 hash function
+def hash_pass(password):
+    hash_obj = hashlib.sha1(password.encode('utf-8')).digest()
+    hash_password = hashlib.sha1(hash_obj).hexdigest()
+    return hash_password
+
+
+# Generates a JSON Web Token for a given username
+def generate_token(username):
+    payload = {
+        'iss': ISSUER,  # Issuer
+        'sub': username,  # Subject
+        'iat': time.time(),  # Issued at
+        'exp': time.time() + EXP_TIME  # Expiration time
+    }
+
+    token = jwt.encode(payload, SECRET_KEY, algorithm='HS256')
+    return token
+
+
+# Determines if a token is valid for a given username
+def is_valid_token(token, username):
```

Comment:
http://flask.pocoo.org/docs/0.12/quickstart/#sessions 

`Flask` sessions api. We might not need to store the token on mongo. This approach might be better because we don't have to access mongo in order to check if a user is logged in.